### PR TITLE
[Dune] Backfill Historical Records

### PR DESCRIPTION
**Pull request by bh2smith**
Adding dependency on dune-client to assist with fetching data for back filling. We introduce a script backfill that works as follows:

```sh
cd internal_transfers/actions
npm i
npm run build
cp .env.sample .env <--- Fill out your credentials
source .env
node out/scripts/backfill.js --dateFrom=2023-01-01 --dateTo=2023-01-02
```

TODO 

- Make a Docker image. (follow up)
- [Optional] Trigger manual run of tenderly from API.

### CODE REVIEW

File: internal_transfers/actions/src/pipeline.ts [hunk 1]
```diff
@@ -95,6 +95,7 @@ export async function internalizedTokenImbalance(
       if (reFetchedReceipt.logs.length === 0) {
         settlementSimulations = null;
       } else {
+        // TODO - TENDERLY ALERT?!
```
Comment (bh2smith) [inline internal_transfers/actions/src/pipeline.ts:98]: This will be done in a follow up PR.

File: internal_transfers/actions/scripts/backfill.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+if (!NODE_URL) {
+  throw new Error("Missing NODE_URL");
+}
+
+let simulator: TransactionSimulator = new EnsoSimulator(
+  "http://127.0.0.1:8080/api/v1/simulate"
+);
+if (
+  ![TENDERLY_ACCESS_KEY, TENDERLY_USER, TENDERLY_PROJECT].includes(undefined)
```
Comment (alfetopito) [inline internal_transfers/actions/scripts/backfill.ts]: Do you mean to say, if any of those are undefined?

If that's the case, I suggest instead:

```suggestion
  TENDERLY_ACCESS_KEY && TENDERLY_USER && TENDERLY_PROJECT
```

Way clearer IMO

File: internal_transfers/actions/scripts/backfill.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+) {
+  console.log("Using Tenderly Simulator");
+  simulator = new TenderlySimulator(
+    TENDERLY_USER!,
+    TENDERLY_PROJECT!,
+    TENDERLY_ACCESS_KEY!
```
Comment (alfetopito) [inline internal_transfers/actions/scripts/backfill.ts]: You won't need `!` if you use the `if` condition as suggested (I think)

File: internal_transfers/actions/scripts/backfill.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+if (!DB_URL) {
+  throw new Error("Missing DB_URL");
+}
+if (!NODE_URL) {
+  throw new Error("Missing NODE_URL");
+}
+
+let simulator: TransactionSimulator = new EnsoSimulator(
+  "http://127.0.0.1:8080/api/v1/simulate"
+);
```
Comment (alfetopito) [inline internal_transfers/actions/scripts/backfill.ts]: Is this an expensive operation?

Since it might be replaced by tenderly if available, I would do it in the `else` condition

File: internal_transfers/actions/scripts/backfill.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  dateTo: String,
+});
+backFillTokenImbalances(
+  args.dateFrom,
+  args.dateTo,
+  DB_URL,
+  NODE_URL,
+  DUNE_API_KEY,
+  simulator
+).then((_) => console.log("Execution Complete"));
```
Comment (alfetopito) [inline internal_transfers/actions/scripts/backfill.ts]: If you don't care about the promise return value, you just not pass any parameter

```suggestion
).then(() => console.log("Execution Complete"));
```

Unless the linter is complaining about it?

File: internal_transfers/actions/src/dune.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+export async function getSampleSet(
+  dune: DuneClient,
+  // YYYY-MM-DD
+  dateFrom: string,
+  dateTo: string
+): Promise<MinimalTxData[]> {
+  const historicalTransactionData = await dune
+    // ~ 7 minutes / day with Tenderly
+    // ~ 39 minutes / day with Enso
+    .refresh(2492342, [
```
Comment (alfetopito) [inline internal_transfers/actions/src/dune.ts]: Is this magic number related to 7 or 39 minutes?
I'm confused.
Can you add it as a const for making it clearer?

File: internal_transfers/actions/src/dune.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+  return historicalTransactionData.rows.map((row: Record<string, any>) => {
+    const { blockNumber, from, hash, logs } = row;
+    if (!isEventLogArray(logs)) {
+      throw Error(`Invalid Row ${JSON.stringify(row)}`);
+    }
+    return {
+      blockNumber: parseInt(blockNumber),
+      from: from,
+      hash: hash,
```
Comment (alfetopito) [inline internal_transfers/actions/src/dune.ts]:
```suggestion
      from,
      hash,
```

File: internal_transfers/actions/scripts/backfill.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    TENDERLY_USER!,
+    TENDERLY_PROJECT!,
+    TENDERLY_ACCESS_KEY!
+  );
+}
+
+export const args = parse<RuntimeArgs>({
+  dateFrom: String,
+  dateTo: String,
+});
```
Comment (alfetopito) [inline internal_transfers/actions/scripts/backfill.ts:51]: Nitpick: do you care about validating these dates match the format you want?

File: internal_transfers/actions/scripts/backfill.ts (hunk 2 above)
Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts]: do you mean if NOT that.

File: internal_transfers/actions/scripts/backfill.ts (hunk 4 above)
Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts]: Yea ok! Good idea

File: internal_transfers/actions/src/dune.ts (hunk 6 above)
Comment (bh2smith) [reply internal_transfers/actions/src/dune.ts]: Will make QUERY_ID a constant. the 7 and 39 minutes are for different simulators. This comment about how long it takes doesn't really belong here.

File: internal_transfers/actions/scripts/backfill.ts (hunk 8 above)
Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts:51]: no. not really, but I am keen on suggestions.

File: internal_transfers/actions/scripts/backfill.ts (hunk 2 above)
Comment (alfetopito) [reply internal_transfers/actions/scripts/backfill.ts]: You want to enter the `if` only if those evn vars of those are set, no?

File: internal_transfers/actions/scripts/backfill.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  DB_URL,
+  NODE_URL,
+  TENDERLY_ACCESS_KEY,
+  TENDERLY_USER,
+  TENDERLY_PROJECT,
+} = process.env;
+
+if (!DUNE_API_KEY) {
+  throw new Error("Missing DUNE_API_KEY");
+}
```
Comment (bh2smith) [inline internal_transfers/actions/scripts/backfill.ts:27]: What about these @alfetopito --- any way to make this less lengthy?

Comment (alfetopito) [reply internal_transfers/actions/scripts/backfill.ts:27]: You could use something like https://www.npmjs.com/package/env-schema to validate both the keys and their contents, but it might be a bit overkill.

File: internal_transfers/actions/scripts/backfill.ts (hunk 8 above)
Comment (alfetopito) [reply internal_transfers/actions/scripts/backfill.ts:51]: You could do everyone's favorite pattern matching tool: regex!
Or a validation lib like https://www.npmjs.com/package/env-schema but I think that's overkill

File: internal_transfers/actions/scripts/backfill.ts (hunk 9 above)
Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts:27]: Fair enough. Once I make a second script (coming soon), I will move the env parsing to a common/shared file and at least only have to do this stuff once.

File: internal_transfers/actions/scripts/backfill.ts (hunk 8 above)
Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts:51]: meh... maybe I'll make a "good first issue" out of it

Comment (bh2smith) [reply internal_transfers/actions/scripts/backfill.ts:51]: https://github.com/cowprotocol/solver-rewards/issues/289

File: internal_transfers/actions/src/dune.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    .refresh(TX_DATA_QUERY_ID, [
+      QueryParameter.date("DateFrom", `${dateFrom} 00:00:00`),
+      QueryParameter.date("DateTo", `${dateTo} 00:00:00`),
+    ])
+    .then((executionResult) => executionResult.result);
+  if (!historicalTransactionData) {
+    throw new Error("Failure");
+  }
+
+  return historicalTransactionData!.rows.map((row: Record<string, any>) => {
```
Comment (alfetopito) [inline internal_transfers/actions/src/dune.ts]: You also don't need `!` here as you throw an error when it's `falsy`

**Review by alfetopito — APPROVED** [review]

File: internal_transfers/actions/src/dune.ts (hunk 10 above)
Comment (bh2smith) [reply internal_transfers/actions/src/dune.ts]: oh ya... Would it also be ok to remove the error throwing and expect the ! to throw instead? Or is it better to handle the error as is already done?

Comment (bh2smith) [reply internal_transfers/actions/src/dune.ts]:
```suggestion
  return historicalTransactionData.rows.map((row: Record<string, any>) => {
```

Comment (alfetopito) [reply internal_transfers/actions/src/dune.ts]: I prefer the explicit throw rather than relying on the a runtime error.
After all, `!` operator is telling TS you know what you are doing and you are sure this value is not falsy
